File: dev/src/HAL/client/plugins/complex.rb
require 'json'

class Coord
  attr_reader :x, :y, :z
  def initialize(x,y,z)
    @x = x
    @y = y
    @z = z
  end
  
  def as_json(option={})
    {
      x: @x,
      y: @y,
      z: @z,
    }
  end

  def to_json(*options)
    as_json(*options).to_json(*options)
  end
end

class Cube
  attr_reader :name, :point

  def initialize 
    @point = Coord.new(0,0,0)
    @name = "cube"
  end

  def as_json(option={})
    {
      name: @name,
      point: @point
    }
  end

  def to_json(*options)
    as_json(*options).to_json(*options)
  end
end

cube = Cube.new

print cube.to_json

File: dev/src/HAL/client/examples/simple_complex/plugins/complex.py
import json

class Point(object):
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

class Cube(object):
    def __init__(self):
        self.point = Point(0, 0, 0)
        self.name = "cube"

def obj_to_dict(obj):
   return obj.__dict__

cube = Cube()

print(json.dumps(cube.__dict__, default=obj_to_dict))
